driver/src/price_estimation/dexag/api.rs

```diff
@@ -0,0 +1,167 @@
+#![allow(dead_code)]
+
+use anyhow::{anyhow, Context, Result};
+use ethcontract::Address;
+use isahc::prelude::*;
+use serde::{de::Error as _, Deserialize, Deserializer};
+use serde_with::rust::display_fromstr;
+use std::time::Duration;
+use url::Url;
+
+#[cfg_attr(test, mockall::automock)]
+/// Parts of the dex.ag api.
+pub trait DexagApi {
+    /// https://docs.dex.ag/api/tokens
+    fn get_token_list(&self) -> Result<Vec<Token>>;
+    /// https://docs.dex.ag/api/price
+    fn get_price(&self, from: &Token, to: &Token, amount: EitherAmount) -> Result<Price>;
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct Token {
+    pub name: String,
+    pub symbol: String,
+    #[serde(default)]
+    #[serde(deserialize_with = "deserialize_address")]
+    pub address: Option<Address>,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct Price {
+    #[serde(with = "display_fromstr")]
+    pub price: f64,
+}
+
+fn deserialize_address<'de, D>(deserializer: D) -> Result<Option<Address>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    fn hex_string_to_address(string: &str) -> Result<Address> {
+        let prefix = "0x";
+        if !string.starts_with(prefix) {
+            return Err(anyhow!("does not start with {}", prefix));
+        }
+        Ok(string[2..].parse()?)
+    }
+
+    let string: Option<String> = Option::deserialize(deserializer)?;
+    Ok(if let Some(string) = string {
+        let address = hex_string_to_address(&string)
+            .with_context(|| format!("failed to parse address \"{}\"", string))
+            .map_err(D::Error::custom)?;
+        Some(address)
+    } else {
+        None
+    })
+}
+
+pub enum EitherAmount {
+    From(f64),
+    To(f64),
+}
+
+#[derive(Debug)]
+pub struct DexagApiImpl {
+    base_url: Url,
+    client: HttpClient,
+}
+
+pub const DEFAULT_BASE_URL: &str = "https://api-v2.dex.ag";
+pub const DEFAULT_TIMEOUT: Duration = Duration::from_secs(10);
+
+impl DexagApiImpl {
+    pub fn new() -> Result<Self> {
+        Self::custom(DEFAULT_BASE_URL, DEFAULT_TIMEOUT)
+    }
+
+    pub fn custom(base_url: &str, timeout: Duration) -> Result<Self> {
+        let client = HttpClient::builder()
```

Comment:
Would be nice if this could use #596, but I guess it will go in first. But then #596 should update this callsite as well.